Refactor coordinates

tomc271 opened the pull request:
Encapsulate all private members of Coordinates class. 
New classes for metric tensor, Christoffel symbols.

tomchapman committed:
Make ShiftTorsion and IntShiftTorsion private.

tomchapman committed:
Clang-Tidy: Pass by value and use std::move().

tomchapman committed:
Rename calculateGeometry() to communicateAndCheckMeshSpacing().

tomchapman committed:
Remove comments referring to obsolete method Coordinates::calculateGeometry().

tomchapman committed:
Extract class ChristoffelSymbols.

tomchapman committed:
Make 'function' parameter to map() and applyToComponents() methods a reference.

tomchapman committed:
Make parameters to metric tensor setters const references.

tomchapman committed:
Rename getters (and private fields) for metric tensor components.

tomchapman committed:
Clang-Tidy: Pass by value and use std::move().

tomchapman committed:
Error handling: null check in Mesh::sourceHasYBoundaryGuards().

tomchapman committed:
Add a constructor for ChristoffelSymbols that takes no Mesh (because FakeMesh has no 'source', which causes a crash when the constructor leads to initialisation of FieldMetric objects).

tomchapman committed:
Clang-Tidy: Pass by value and use std::move().

tomchapman committed:
Clang-Tidy: Single-argument constructors must be marked explicit to avoid unintentional implicit conversions.

tomchapman committed:
Remove redundant 'mesh' parameter from Geometry constructor.

tomchapman committed:
Formatting.

tomchapman committed:
Remove Geometry class, as now redundant.

tomchapman committed:
Corrected comment.

tomchapman committed:
Need to call setter directly on contravariantMetricTensor and covariantMetricTensor in Coordinates::setBoundaryCells method, to avoid recalculation of the christoffel symbols, which would (at this stage) involve field data at different locations (specifically the metric tensor and jacobian).

tomchapman committed:
Ensure covariantMetricTensor is set.

tomchapman committed:
Don't pass mesh to ChristoffelSymbols constructor in the Coordinates constructor that is used with FakeMesh.

tomchapman committed:
Null check differential_operators in ChristoffelSymbols constructor.

tomchapman committed:
Ensure covariantMetricTensor is set.

tomchapman committed:
Pass (a reference to) Coordinates to CalculateChristoffelSymbols(), rather than the metric tensor, dx, dy, etc.

tomchapman committed:
Use same pattern as for invSg for lazy evaluation of Coordinates::christoffel_symbols.

tomchapman committed:
Prefer const.

tomchapman committed:
Remove unused ChristoffelSymbols constructor.

tomchapman committed:
G1, G2, G3 not part of christoffel symbols.

tomchapman committed:
interpolateAndExtrapolate G1, G2, G3.

tomchapman committed:
christoffel_symbols() return type should be reference.

tomchapman committed:
Tidy.

tomchapman committed:
Refactor.

tomchapman committed:
Formatting.

tomchapman committed:
Refactor method 'setBoundaryCells' (remove getting dx and dy from the input source).

tomchapman committed:
Renaming to distinguish between 'options' variables.

tomchapman committed:
Make CalculateChristoffelSymbols() method just another ChristoffelSymbols constructor.

tomchapman committed:
Extract variables for metric tensor components.

tomchapman committed:
Simplify case switching (reduce nesting) in Coordinates::setParallelTransform().

tomchapman committed:
Extract method getDzFromOptionsFile().

tomchapman committed:
Remove 'getAtLoc' function overload that updates an in/out variable, and the remaining two uses of it.

tomchapman committed:
Remove redundant initialisation of IntShiftTorsion, as it will already have been set.

tomchapman committed:
Remove method 'getUnalignedAtLocationAndFillGuards'.

tomchapman committed:
Use member variables 'localmesh' and 'nz' in methods, rather than passing it in as a parameter or re-fetching.

tomchapman committed:
Rename private field J_ to jacobian_cache.

tomchapman committed:
Don't recalculate Jacobian and Bxy if read from file.

tomchapman committed:
Don't recalculate Bxy if read from file.

tomchapman committed:
DRY - use existing method recalculateBxy().

tomchapman committed:
Make recalculateBxy() const.

tomchapman committed:
No need to recalculate Bxy twice.

tomchapman committed:
Use extrapolate_x, extrapolate_y options if set.
(So now we can directly replace jacobian() with recalculateJacobian(); no longer a special case).

tomchapman committed:
Reduce code duplication by moving call to interpolateAndExtrapolate() outside of if-else statement.

tomchapman committed:
Refactor to use sourceHasVar() and getAtLoc() for IntShiftTorsion, as they are for the other variables.

tomchapman committed:
No need to assign to a local variable before setting property.

tomchapman committed:
Make jacobian_cache a std::unique_ptr.
Set after comparing calculated and loaded values, otherwise there will be no difference.
Rename Jcalc to J_from_file.

tomchapman committed:
Bug fix: update existing value of jacobian when setting for specific x, y indices.

tomchapman committed:
Remove ShiftTorsion setter (to avoid problems with it not being initialised fully (no mesh)).

tomchapman committed:
Revert extracted functions on DifferentialOperators for those that are only used by Coordinates class (i.e. put back dependency on bout/derivs).

tomchapman committed:
Extract class GValues to encapsulate G1, G2, G3 (needs renaming when we know what the name should be) and implement with same pattern as Christoffel symbols.

tomchapman committed:
Extract methods Coordinates::communicateGValues() and Coordinates::extrapolateGValues().
Make GValues constructor take a reference to Coordinates (so that it can call methods on the Coordinates class (DDX, DDY, DDZ, communicate).
Make communicate() a (public) method of Coordinates so that it can be called by the GValues constructor.
This means fixZShiftGuards() also has to be a (private) method of Coordinates.

tomchapman committed:
Remove redundant MetricTensor::Allocate() method.

tomchapman committed:
Fix additional strings after renaming "MetricTensor::CalculateOppositeRepresentation" to "MetricTensor::inverse".

tomchapman committed:
Define two derived classes CovariantMetricTensor and CovariantMetricTensor that inherit from MetricTensor.

tomchapman committed:
Fully revert extracting differential operator methods from Coordinates class onto DifferentialOperators.

tomchapman committed:
Move trivial getters and setters of Coordinates class to header file.

tomchapman committed:
Move trivial getters and setters of ChristoffelSymbols class to header file.

tomchapman committed:
Move trivial getters and setters of GValues class to header file.

tomchapman committed:
Move trivial getters and setters (and setLocation() method) of MetricTensor class to header file.

bendudson commented on examples/2Dturbulence_multigrid/esel.cxx:
```diff
@@ -69,23 +69,25 @@ class ESEL : public PhysicsModel {
-    coord->geometry();
+    coord->setBxy(1);
+
+    MetricTensor::FieldMetric g11, g22, g33, g12, g13, g23;
+    g11 = 1.0;
+    g22 = 1.0;
+    g33 = 1.0;
+    g12 = 0.0;
+    g13 = 0.0;
+    g23 = 0.0;
+    coord->setContravariantMetricTensor(ContravariantMetricTensor(g11, g22, g33, g12, g13, g23));
+
+    MetricTensor::FieldMetric g_11, g_22, g_33, g_12, g_13, g_23;
+    g_11 = 1.0;
+    g_22 = 1.0;
+    g_33 = 1.0;
+    g_12 = 0.0;
+    g_13 = 0.0;
+    g_23 = 0.0;
+    coord->setCovariantMetricTensor(CovariantMetricTensor(g_11, g_22, g_33, g_12, g_13, g_23));
```
```suggestion
    coord->setContravariantMetricTensor(ContravariantMetricTensor(1.0, 1.0, 1.0, 0.0, 0.0, 0.0));
    coord->setCovariantMetricTensor(CovariantMetricTensor(1.0, 1.0, 1.0, 0.0, 0.0, 0.0));
```

tomc271 commented on examples/6field-simple/elm_6f.cxx:
```diff
@@ -1051,24 +1051,26 @@ class Elm_6f : public PhysicsModel {
+    g_12 = Btxy * hthe * I * Rxy / Bpxy;
+    g_13 = I * Rxy * Rxy;
+    g_23 = Btxy * hthe * Rxy / Bpxy;
+    coord->setCovariantMetricTensor(CovariantMetricTensor(g_11, g_22, g_33, g_12, g_13, g_23));
```
It will do, yes. The Coordinates constructor is such a function -- just use that?

ZedThree replied:
It doesn't look like there is currently a constructor that takes more metrics. We probably want a method `setBothMetricTensors(ContravariantMetricTensor, CovariantMetricTensor)`.

tomc271 replied:
I was thinking of this constructor: 
`  Coordinates(Mesh* mesh, FieldMetric dx, FieldMetric dy, FieldMetric dz, FieldMetric J,
              FieldMetric Bxy, const FieldMetric& g11, const FieldMetric& g22,
              const FieldMetric& g33, const FieldMetric& g12, const FieldMetric& g13,
              const FieldMetric& g23, const FieldMetric& g_11, const FieldMetric& g_22,
              const FieldMetric& g_33, const FieldMetric& g_12, const FieldMetric& g_13,
              const FieldMetric& g_23, FieldMetric ShiftTorsion,
              FieldMetric IntShiftTorsion);`

bendudson commented:
Thanks @tomc271 ! This is quite an epic amount of work.

tomc271 committed:
No need to pre-assign values to explanatory variables before using to construct a MetricTensor.

Co-authored-by: Ben Dudson <dudson2@llnl.gov>

ZedThree reviewed (commented):
Thanks @tomc271! I've made a bunch of comments on things to work through. A couple of big things though:

- this has diverged from `next` a bit, so that will need merging in and conflicts resolving. There's been some important bug fixes and CI changes in `next` that would be good to get in ASAP
- some of the changes haven't propagated through to some subsystems, such as the various PETSc laplace operators. You will need to either build against PETSc locally to find them, or use some searching to find the remaining places

tomc271 commented on include/bout/mesh.hxx:
```diff
@@ -795,6 +809,17 @@ public:
+  /// Boundary guard cells are set by extrapolating from the grid, like
+  /// 'free_o3' boundary conditions
+  /// Corner guard cells are set to BoutNaN
+  const Field2D interpolateAndExtrapolate(const Field2D& f, CELL_LOC location,
```
I think I moved it during my attempts to reduce coupling. If anything, perhaps it should go in Field2D. (Even better use a creational pattern for constructing a Coordinates object). Anyway, I'll put it back.

tomc271 commented on src/mesh/christoffel_symbols.cxx:
```diff
@@ -0,0 +1,156 @@
+  const auto& g_13 = covariantMetricTensor.g13();
+  const auto& g_23 = covariantMetricTensor.g23();
+
+  G1_11_ = 0.5 * g11 * coordinates.DDX(g_11)
```
Ok, but it originally called `Coordinates::DDX`

tomchapman committed:
Remove obsolete inverse() and jacobian() in boutcpp.pxd.jinja file.

tomchapman committed:
Restore semicolons after macros.

tomchapman committed:
Remove redundant `const`.

tomchapman committed:
Fix MetricTensor::inverse method (it was replacing each metric component with a single scalar value).

tomc271 commented on src/mesh/metricTensor.cxx:
```diff
@@ -0,0 +1,153 @@
+      output_error.write(error_message, i.x(), i.y());
+      throw BoutException(error_message);
+    }
+  }
+
+  BoutReal g_11, g_22, g_33, g_12, g_13, g_23;
+  g_11 = a(0, 0);
+  g_22 = a(1, 1);
+  g_33 = a(2, 2);
+  g_12 = a(0, 1);
+  g_13 = a(0, 2);
+  g_23 = a(1, 2);
```
Slightly concerning that the tests didn't catch this

tomchapman committed:
Restore commented-out code (Maximum error in diagonal/off-diagonal inversion).

tomchapman committed:
Use 'auto' when type name is obvious.

tomchapman committed:
Clang-tidy: "Use auto when initialising with a template cast to avoid duplicating the type name"

tomchapman committed:
Bug fix: Don't specify region when setting contravariant metric tensor in 'interpolateFieldsFromOtherCoordinates' method.

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates

tomchapman committed:
Update petsc cases to use new getters.

tomchapman committed:
Remove calls to geometry().

tomchapman committed:
Set death_test_style to "threadsafe", to prevent timeout.

tomchapman committed:
Add [[maybe_unused]] attribute to 'J' parameter in Coordinates constructor.

tomchapman committed:
Revert expansion of COPY_STRIPE macro.

tomchapman committed:
Remove redundant function 'ChristoffelSymbols::map' (instead just calling ChristoffelSymbols::applyToComponents() directly).

tomchapman committed:
Revert "Minor fixes."

This reverts commit 5d2a9dd10f75601382c85d85eef2d73e419182b1.

(Return value could be either Field2D or Field3D).

tomc271 committed:
Apply clang-format changes

dschwoerer reviewed (commented):
Thanks @tomc271 for this PR. It is however not quite clear to me what the aim of this PR is.

It is changing many files, but I am failing to see the goal of such a change. What are you trying to improve?

dschwoerer commented on examples/laplacexy/alfven-wave/alfven.cxx:
```diff
@@ -188,11 +188,11 @@ class Alfven : public PhysicsModel {
     Bpxy /= Bnorm;
     Btxy /= Bnorm;
-    coord->Bxy /= Bnorm;
+    coord->setBxy(coord->Bxy() / Bnorm);
```
`dx() = ...` is from here: https://github.com/boutproject/BOUT-dev/pull/2873/files/b8a2158beabb206b6b3cb356aad9c2fb507797ce#diff-3dd433883dc95e06b333ad5c044798cba4ed24d5327d83ef730df001975febe0R182

tomc271 replied:
> `dx() = ...` is from here: https://github.com/boutproject/BOUT-dev/pull/2873/files/b8a2158beabb206b6b3cb356aad9c2fb507797ce#diff-3dd433883dc95e06b333ad5c044798cba4ed24d5327d83ef730df001975febe0R182

Good spot, thanks @dschwoerer. Fixed now.

ZedThree commented:
@dschwoerer The aim of this work is to ensure that `Coordinates` is always correct, and that modifications to the metric are correctly and automatically propagated to the derived quantities, as well as enabling lazy evaluation of the metric derivatives and Christoffel symbols. The lazy evaluation will also let us break some painful and annoying dependency cycles.

> changing dx from a Field to a function will be a major pain for porting models, and makes it impossible to write portable code.

> Could coord->dx be a new wrapper class that forwards everything to the real dx?
That would be more work for BOUT++ but would be more API conserving.
As it currently is, I think this might be suitable for v6.0.0 - I am not sure whether we want to have more v5.x.0 releases before that?

We might investigate proxy classes in the future, but they have some quite painful downsides. But yes, this is expected to be a breaking change for v6. I'm not sure there's anything planned for v5.x.

Another planned breaking change is to change `g11` -> `gxx` and so on (in a separate PR). Most of these changes will be fairly straightforward to port in user code, so we should be able to provide an automated tool similar to the `bout-v5-upgrader` scripts.

tomchapman committed:
More updates to examples and tests to use dx setter.

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/refactor-coordinates' into refactor-coordinates

tomchapman committed:
Rename file MetricTensor to use snake case.

tomchapman committed:
Updated another example (gravity_reduced) to use new metric tensor setters.

tomc271 committed:
Apply clang-format changes

tomc271 commented on include/bout/christoffel_symbols.hxx:
```diff
@@ -0,0 +1,93 @@
+                             const FieldMetric& G3_11, const FieldMetric& G3_22,
+                             const FieldMetric& G3_33, const FieldMetric& G3_12,
+                             const FieldMetric& G3_13, const FieldMetric& G3_23) {
+    G1_11_ = G1_11;
```
What naming convention do you tend to use for private fields?

dschwoerer replied:
Here you could use `var_G1_11` or `cache_G1_11`?

tomc271 replied:
The trailing underscore is arguably the most popular and recommended modern convention. It is famously enforced by the Google C++ Style Guide. 
The second common option would be the "m_" prefix.

tomchapman committed:
Remove obsolete comment.

tomchapman committed:
Make jacobian setter take a const reference.

tomc271 commented on tools/pylib/_boutpp_build/helper.cxx.jinja:
```diff
@@ -170,8 +170,8 @@ Mesh * c_get_global_mesh(){
+  coord->setDx(coord->dx() / norm);
+  coord->setDy(coord->dy() / norm);
+  coord->setDz(coord->dz() / norm);
+  coord->communicateAndCheckMeshSpacing();
```
Perhaps -- I don't know. This method is just what is left of Coordinates::geometry() after extracting everything else.

tomchapman committed:
Call communicateAndCheckMeshSpacing() in ChristoffelSymbols constructor (immediately before calling the differential operator functions).

tomc271 commented on src/mesh/coordinates.cxx:
```diff
@@ -932,461 +615,208 @@ const Field2D& Coordinates::zlength() const {
+int Coordinates::communicateAndCheckMeshSpacing() const {
+  TRACE("Coordinates::communicateAndCheckMeshSpacing");
+
+  auto tmp = dx(); // TODO: There must be a better way than this!
+  communicate(tmp, dy(), dz(), g11(), g22(), g33(), g12(), g13(), g23(), g_11(), g_22(),
+              g_33(), g_12(), g_13(), g_23(), J(), Bxy());
```
The 'communicate' method appears to take references (to the variables), which is what the getters return:  
`FieldGroup(Ts&... ts)`
Also all the tests are passing?

dschwoerer commented:
> @dschwoerer The aim of this work is to ensure that `Coordinates` is always correct, and that modifications to the metric are correctly and automatically propagated to the derived quantities, as well as enabling lazy evaluation of the metric derivatives and Christoffel symbols. The lazy evaluation will also let us break some painful and annoying dependency cycles.

Thanks @ZedThree for the motivation. That makes a lot of sense, and certainly worth it to break some code, if needed.

> > changing dx from a Field to a function will be a major pain for porting models, and makes it impossible to write portable code.
> 
> > Could coord->dx be a new wrapper class that forwards everything to the real dx?
> > That would be more work for BOUT++ but would be more API conserving.
> > As it currently is, I think this might be suitable for v6.0.0 - I am not sure whether we want to have more v5.x.0 releases before that?
> 
> We might investigate proxy classes in the future, but they have some quite painful downsides. But yes, this is expected to be a breaking change for v6. I'm not sure there's anything planned for v5.x.

If we do add wrappers, should we not add them now?

I have added a draft here:
https://github.com/boutproject/BOUT-dev/blob/field-tracking-draft/include/bout/coordinates.hxx#L87
It may be nicer to take a lambda in the constructor, rather than providing classes for all cases - but that should be easy enough to change.

> Another planned breaking change is to change `g11` -> `gxx` and so on (in a separate PR). Most of these changes will be fairly straightforward to port in user code, so we should be able to provide an automated tool similar to the `bout-v5-upgrader` scripts.

That should be easy enough to retain backwards API compatibility?

ZedThree commented:
> If we do add wrappers, should we not add them now?

I want to make sure we get the features in this PR correct first, and then look at the wrappers. There's some non-obvious pitfalls that proxy classes can go wrong, so it's probably easier to explore that in a separate PR. It's also worth looking to see if there's some existing C++ libraries for helping out.

> That should be easy enough to retain backwards API compatibility?

It would definitely be good to try and retain backwards compatibility where possible, especially as it would allow us to mark bits as deprecated.

tomchapman committed:
Remove unused #include.

tomchapman committed:
Clang-Tidy: Single-argument constructors must be marked explicit to avoid unintentional implicit conversions.

tomchapman committed:
Remove unused #include directives.

tomchapman committed:
Prefer const.

tomchapman committed:
Add [[maybe_unused]] attributes.

tomchapman committed:
More unused #include directives.

tomchapman committed:
Clang-Tidy: The 'empty' method should be used to check for emptiness instead of 'length'.

tomchapman committed:
More [[maybe_unused]] attributes.

tomchapman committed:
Clang-Tidy: Method can be made static.

tomchapman committed:
Clang-Tidy: Constructors that are callable with a single argument must be marked explicit to avoid unintentional implicit conversions.

tomchapman committed:
Another unused #include directive.

tomchapman committed:
More [[maybe_unused]] attributes and unused #include directives.

tomchapman committed:
Use std::move() for constructor parameters.

dschwoerer commented:
> I want to make sure we get the features in this PR correct first, and then look at the wrappers. There's some non-obvious pitfalls that proxy classes can go wrong, so it's probably easier to explore that in a separate PR. It's also worth looking to see if there's some existing C++ libraries for helping out.

So this is still WIP?

If you get to the point to look at wrappers, feel free to look at https://github.com/boutproject/BOUT-dev/tree/field-tracking-draft 

> It would definitely be good to try and retain backwards compatibility where possible, especially as it would allow us to mark bits as deprecated.

I think we should care more about this :-)

tomchapman committed:
More unused #include directives.

tomchapman committed:
No need to communicateAndCheckMeshSpacing() after creating Coordinates.

tomchapman committed:
Revert "Move interpolateAndExtrapolate() to Mesh class."

This reverts commit 37c90c5a6d27d15a043362a46f6fd783e5df28a1.

It now has to be a class method, as it's called by other functions that are now member functions.

tomchapman committed:
Fix using FieldMetric = Field3D.

tomchapman committed:
Fix Field3D Coordinates::interpolateAndExtrapolate method.

tomchapman committed:
Add 3D version of setDy() for the BOUT_USE_METRIC_3D case.

tomchapman committed:
d2z can't be const to work with Mesh::get().

tomchapman committed:
Update BOUT_USE_METRIC_3D case to use metric tensor getters.

tomchapman committed:
Return type of Grad_par() and Vpar_Grad_par() should be FieldMetric, not Field2D

(as it could be either Field2D or Field3D).

tomchapman committed:
Add 3D version of setJ() for the BOUT_USE_METRIC_3D case.

tomchapman committed:
Return type of the std::function parameter to MetricTensor::map should be FieldMetric, not Field2D

(as it could be either Field2D or Field3D).

tomchapman committed:
Return type of Mesh::get() should be FieldMetric, not Field2D

(as it could be either Field2D or Field3D).

tomchapman committed:
Fix updating of Bxy by splitParallelSlices().

tomchapman committed:
Fix updating of Bxy yup and ydown.

tomchapman committed:
Fix updating of Bxy by splitParallelSlices().

tomchapman committed:
Fix updating of Bxy yup and ydown.

tomchapman committed:
G1, G2, G3 are now calculated in the GValues constructor, so no need to set to arbitrary values in the tests.

Also the constructor calls the differential operators, which (in the BOUT_USE_METRIC_3D case) require that the ParallelTransform has already been set.

tomchapman committed:
Remove call to obsolete method coords._setmembers().

tomchapman committed:
Refactoring to DRY up Mesh::createDefaultCoordinates().

tomchapman committed:
Add new coordinates object to coords_map before calling recalculateAndReset().

Otherwise calculation of the Christoffel symbols and metric tensor will cause Mesh::getCoordinates() to be called (and fail).

(This is because of the circular dependency between Mesh and Coordinates).

tomchapman committed:
Improve exception handling - ensure that transform is not null before attempting to use it in Coordinates::DDY().

tomchapman committed:
Ensure transform is set even when Coordinates is instantiated with the constructor that takes no Options

(Use the basic type, ParallelTransformIdentity, in this case).

tomchapman committed:
Bug fix: Mesh::get() should return a Field3D in the BOUT_USE_METRIC_3D case.

tomchapman committed:
Merge branch 'clang-tidy' into refactor-coordinates

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates

tomchapman committed:
Merge branch 'fix-for-3d' into refactor-coordinates

tomchapman committed:
Merge branch 'fix-for-3d' into refactor-coordinates

tomchapman committed:
Need to ensure parallel transform is set before differential operators are used
(and setParallelTransform() requires that dz is already set!)

tomchapman committed:
Don't setLocation() of dz at this point, to avoid ASSERT1_FIELDS_COMPATIBLE(lhs, rhs) failing in Field3D operator/(const Field3D& lhs, const Field3D& rhs).

tomchapman committed:
Reorder to put the variable closer to its point of use.

tomchapman committed:
Fix type of dJ_dy in LaplaceHypre3d::updateMatrix3D() (3D not 2D).

tomc271 committed:
Apply clang-format changes

tomchapman committed:
Fix test-laplacexz: Use a non-zero value for off-diagonal components that gives a valid inverse metric (i.e. all diagonal components positive).

This became a problem once setting the contra- or co-variant metric tensor was made to trigger the calculation of the other representation.

tomchapman committed:
Disable "Laplacian solver in 2D (X-Y)" for the BOUT_USE_METRIC_3D case.

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates

dschwoerer commented:
On 3/19/24 16:13, Peter Hill wrote:
> I want to make sure we get the features in this PR correct first, and 
> then look at the wrappers.

So this is still a draft until the wrappers are implemented?

tomchapman committed:
Merge branch 'next' into refactor-coordinates

Commit:
include <functional>

tomchapman committed:
Add Coordinates::setMetricTensor() method for setting both co- and contravariant components.

tomchapman committed:
Call Mesh::communicate() rather than Coordinates::communicate() when multiple parameters are sent.

tomchapman committed:
Replace variadic template with two methods: void communicate(const Field2D& f) and void communicate(const Field3D& f).

tomchapman committed:
Add a version of Mesh::communicate<T>() that takes const-qualified parameters.

tomchapman committed:
Allow FieldGroup to take const-qualified parameters.

Some duplication of methods because there are still cases where it has to take non-const items.

tomchapman committed:
Add a version of Coordinates::communicate<T>() that takes const-qualified parameters.

tomchapman committed:
This fixes the failure of tests `NotConstructableFromInt` and `NotConstructableFromInt`.

tomchapman committed:
Various Clang-Tidy suggestions.

tomchapman committed:
Do the copying in the method that calls communicate(), so that the copies are still in scope when they are used.

tomchapman committed:
Do the copying in the method that calls Mesh::communicate(), so that the copies are still in scope when they are used.

tomchapman committed:
Merge branch 'communicate-const-qualified' into refactor-coordinates

ZedThree commented on src/mesh/mesh.cxx:
```diff
@@ -10,8 +9,6 @@
 #include <bout/boutcomm.hxx>
 #include <bout/output.hxx>
 
-#include "impls/bout/boutmesh.hxx"
```
I think this is probably what's breaking the static builds as the `RegisterMesh<BoutMesh> registermeshbout{"bout"}` variable won't be seen anywhere.

ZedThree commented on src/field/vector2d.cxx:
```diff
@@ -67,7 +65,9 @@ void Vector2D::toCovariant() {
     if (location == CELL_VSHIFT) {
-      Coordinates *metric_x, *metric_y, *metric_z;
+      Coordinates* metric_x;
+      [[maybe_unused]] Coordinates* metric_y;
```
Why is this `[[maybe_unused]]`? Marking a local variable `[[maybe_unused]]` should set off alarm bells in your head -- look at the diff of L91-94 below!

ZedThree commented on src/mesh/coordinates.cxx:
```diff
@@ -8,43 +8,98 @@
-  t.getMesh()->wait(h);
-  h = t.getMesh()->sendX(g);
-  t.getMesh()->wait(h);
+void Coordinates::communicate(const Field2D& f) {
+  auto f_copy =
+      f; // Copy value to remove const qualifier (because std::vector can't take cv-qualified items)
+  FieldGroup g(f_copy);
```
This just communicates the local copy, so doesn't do what you think it does.

tomchapman committed:
Fix copy-paste errors introduced during refactoring.

tomchapman committed:
Restore #include "impls/bout/boutmesh.hxx"

(removed in commit 62dcf21e68535cb755e3ad88a86f533863bdc95f)

tomchapman committed:
Add MetricTensor::communicate() method, since need to pass a non-const reference to the actual data to Mesh::communicate(), so that it can modify it.

tomchapman committed:
Add ChristoffelSymbols::communicate() method.

tomchapman committed:
Pass (const refs to) the actual d1_dx, d1_dy, d1_dz data to localmesh->communicate(), so that it will be able to modify it.

tomchapman committed:
Add GValues::communicate() method.

tomchapman committed:
Remove redundant methods Coordinates::communicateChristoffelSymbolTerms() and Coordinates::communicateGValues().

tomchapman committed:
Coordinates::communicate() method needs to take a non-const parameter to be able to modify it.

tomchapman committed:
Revert "More [[maybe_unused]] attributes."

This reverts commit c7b70a9fb8af0d11368d019f2bc328260c366ff4.

tomchapman committed:
Revert "Add [[maybe_unused]] attributes."

This reverts commit 0235449bca00cde8b922bb20e2620af4623b588e.

tomchapman committed:
Revert "Prefer const."

This reverts commit 77231a6dece68e55609be962eabf7329a61b928f.

tomchapman committed:
Merge branch 'communicate-fix' into fx01

tomchapman committed:
Add test CheckCovariantCalculatedFromContravariant to CoordinatesTest suite.

tomchapman committed:
Add test CheckContravariantCalculatedFromCovariant to CoordinatesTest suite.

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates

tomchapman committed:
Use new method Coordinates::setMetricTensor() when manually setting both co- and contravariant components in the tests and examples.

tomchapman committed:
Define local variables on assignment (and using auto).

tomchapman committed:
Prefer const.

tomchapman committed:
Fix copy-paste errors introduced during refactoring (more).

tomchapman committed:
Define local variables on assignment (and using auto).

Other clang-tidy suggestions.

tomchapman committed:
Revert "Clang-Tidy: Constructors that are callable with a single argument must be marked explicit to avoid unintentional implicit conversions."

This reverts commit befabd5f73737432800d7d0d460b76f7c32762f9.

tomchapman committed:
Revert "Use std::move() for constructor parameters."

This reverts commit ac9cad6d105786761334a44c253919f00b7e8e5f.

tomchapman committed:
Revert "Another unused #include directive."

This reverts commit b2a3dadc623ae5e8da47a2bb868a18c8bbf07dc9.

tomchapman committed:
Revert "Clang-Tidy: Single-argument constructors must be marked explicit to avoid unintentional implicit conversions."

This reverts commit 9adbfda5e3cd43c4271278e7b74606541539ecdc.

tomchapman committed:
Revert "Remove redundant semicolons."

This reverts commit 40de3eef62122d25ef2ec63138dbc2ba3a549760.

tomchapman committed:
Revert "More [[maybe_unused]] attributes."

This reverts commit e37d6fe210c9e036f45ce7fc201391a7b0e7bbea.

tomchapman committed:
Revert "More unused #include directives."

This reverts commit b719ad10fc93516897b876a03d601644eaea261c.

tomchapman committed:
Revert "Clang-Tidy: The 'empty' method should be used to check for emptiness instead of 'length'."

This reverts commit 271825057a28ce430d35c0558486b90664a2ee43.

tomc271 committed:
Apply clang-format changes

tomchapman committed:
Restore semicolons after macros.

tomchapman committed:
Revert accidental renaming in docs.

tomchapman committed:
Revert "Clang-Tidy: Method can be made static."

This reverts commit 07078edd399b49e66f86a4b5b746a3d3c575979a.

tomchapman committed:
Revert "More unused #include directives."

This reverts commit 0dfafd42d10259079f32be638898ab6123664d33.

tomchapman committed:
Clang-Tidy: Static member accessed through instance.

tomchapman committed:
Clang-Tidy: variable can be declared 'const'.

tomc271 committed:
Apply clang-format changes

tomchapman committed:
Rename setBoundaryCells() to readFromMesh(), and interpolateFieldsFromOtherCoordinates() to interpolateFromCoordinates().

tomchapman committed:
x, y, z are accessed through getters on BoundaryRegionPar.

tomchapman committed:
Remove const qualifier from declaration of temporary copy of Bxy.

tomchapman committed:
Bug fix: can't assign to a const ref - Use a temporary variable with mesh->get().

tomchapman committed:
Bug fix: can't assign to a const ref - Use a temporary variable with mesh->get().

dschwoerer commented:
Just to clarify: This is still WIP?
Could you please give a heads up when this is ready for review?

tomchapman committed:
Define local variables on assignment (and using auto).

tomchapman committed:
Explicity setting the jacobian has to come after setting the metric tensor
because it will invalidate the (cached) jacobian, which will then be recalculated from the metric tensor).

tomc271 committed:
Apply clang-format changes

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates

Commit:
Add overload getters for dx, dy, and dz, that give the value at a particular x, y (or x, y, z) position.

tomchapman committed:
Use overloaded getters for dx, dy, and dz.

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates

tomchapman committed:
Add overloaded getters for d1_dx, d1_dy, and d1_dz, that give the value at a particular x, y (or x, y, z) position.

tomchapman committed:
Add overloaded getters for metric tensor components, that give the value at a particular x, y (or x, y, z) position.

tomchapman committed:
Use overloaded getters for metric tensor components.

tomchapman committed:
Getters for metric tensor components given (x, y) indices.

tomchapman committed:
The (x, y) indexing methods return a pointer when BOUT_USE_METRIC_3D and a reference otherwise).

tomchapman committed:
The (x, y) indexing methods return a pointer when BOUT_USE_METRIC_3D and a reference otherwise).

tomchapman committed:
Disable metric tensor getter with (x, y) arguments when BOUT_USE_METRIC_3D.

tomchapman committed:
Use overloaded getters for metric tensor components.

tomchapman committed:
Improve readability of preprocessor directive

tomchapman committed:
Add overloaded getter for IntShiftTorsion that gives the value at a particular x, y (or x, y, z) position.

tomchapman committed:
Add overloaded getter for J() that gives the value at a particular x, y (or x, y, z) position

tomchapman committed:
Add overloaded getter for zlength() that gives the value at a particular x, y (or x, y, z) position

tomchapman committed:
Add overloaded getters for G1, G2, G3, that give the value at a particular x, y (or x, y, z) position

tomc271 committed:
Apply clang-format changes

ZedThree committed:
Revert "x, y, z are accessed through getters on BoundaryRegionPar."

This reverts commit 541ff17b3e424f1eca57bf672ec650d274333218.

Issue is fixed in separate PR

ZedThree committed:
Fix clang-tidy const-correctness warnings

ZedThree committed:
Merge branch 'next' into refactor-coordinates

* next: (72 commits)
  Make function signature match implementation in example
  Fix some clang-tidy issues in dalf3
  Apply suggestions from code review
  Apply clang-format changes
  prefer `auto *` for pointer types
  Fix: preserve regionID
  Fix exception message
  Ensure pointer is checked before dereferencing
  Try to do the right thing for linking netCDF::netcdf
  Add LC gitlab CI for GPU build/run tests
  BoutMask: Add non-const operator[](Ind3D)
  Use consistently signed char
  Apply clang-format changes
  naulin laplace: Acceptance tolerances after maxits
  CI: Remove test of deleted example
  CI: Bump all ubuntu images
  Read 2D variables into Field3D
  Fix "call to deleted constructor of 'Options' [clang-diagnostic-error]"
  Apply clang-format changes
  Revert change to reference for 'datadir'.
  ...

ZedThree committed:
Rename metric tensor component members

ZedThree committed:
Use `std::array` instead of C array

ZedThree committed:
Fix some clang-tidy warnings in metric_tensor

ZedThree committed:
Reduce duplication in metric tensor between 2D/3D metric a little

ZedThree committed:
Inline metric tensor function application `map`

ZedThree committed:
Use free-function derivatives in metric tensor classes

ZedThree committed:
Remove some unused ctors from metric tensor classes

ZedThree committed:
Ensure metric derivatives are always communicated

ZedThree committed:
Remove `MetricTensor::setLocation`

Location will already be set from initialising using `emptyFrom`

ZedThree committed:
Pull out duplicated region name

ZedThree committed:
Remove specialised `setDy`, `setJ` overloads

ZedThree committed:
Ensure grid spacings, Bxy, J, are always communicated

ZedThree committed:
Ensure metric terms are always communicated

ZedThree committed:
Simplify `recalculateJacobian` a little

ZedThree committed:
Inline some single-use functions

ZedThree committed:
Eliminate some unnecessary local variables

ZedThree committed:
Fix wrong type in MMS/gbs

tomchapman committed:
Set coordinates on FakeMesh

otherwise when Coordinates setters call mesh->communicate()
and that calls calcParallelSlices()
and that calls getCoordinates()
the coordinates field will be a null pointer.

tomchapman committed:
Add optional boolean parameter `communicate` to Coordinates setters and MetricTensor inverse() method

Add method Coordinates::communicateMetricTensor() and call after constructing Coordinates

to avoid circular dependency (between Mesh and Coordinates) as a result of the setter methods calling communicate()

tomchapman committed:
Add Coordinates::communicateDz() and call after constructing Coordinates

to prevent getUniform(coords->zlength()) failing because the field is not const.

tomchapman committed:
Fix bug in test_snb

Update existing dy and J fields, to avoid invalid values in the guard cells.

tomc271 committed:
Apply clang-format changes

bendudson committed:
Merge pull request #3056 from boutproject/fix-after-reactoring-communicate

Fixes to address failing tests after changes to refactor-coordinates

tomchapman committed:
Merge remote-tracking branch 'remotes/boutproject/next' into refactor-coordinates-with-next-merged-in

tomchapman committed:
Fix merge

Include changes from commit: "Return `std::optional` from `invert3x3`"
SHA-1: c42dc24910ef7a4131af48f1e94398c0de1aaab5
which was not merged due to splitting files.

tomchapman committed:
Formatting (whitespace changes only)

tomchapman committed:
Fix for failing fci tests. Need to use Coordinates::DDY() not bout::derivatives::index::DDY()
as the former checks canToFromFieldAligned().

tomchapman committed:
Revert "Use free-function derivatives in metric tensor classes"

This reverts commit e5010e466102609da4cc0fc4b045585ab4b7e992.

Fix for failing fci tests.

tomc271 committed:
Apply black changes

dschwoerer commented:
It was my understanding that you are going to re-enable to do `coords->dy[i]`, instead of `coords->dy()[i]` - do you still plan to do that, or is that not planned?

That would not only make it mostly backwards compatible, but it would also reduce the diff hugely, as most of the BOUT++ code base would not need to be touched.

tomc271 commented:
> It was my understanding that you are going to re-enable to do `coords->dy[i]`, instead of `coords->dy()[i]` - do you still plan to do that, or is that not planned?
> 
> That would not only make it mostly backwards compatible, but it would also reduce the diff hugely, as most of the BOUT++ code base would not need to be touched.

[This summarises our thinking](https://gemini.google.com/share/843245d3d8c9)

dschwoerer commented:
> [This summarises our thinking](https://gemini.google.com/share/843245d3d8c9)

I get the error that this link is invalid or has been deleted.

dschwoerer commented:
> > [This summarises our thinking](https://gemini.google.com/share/843245d3d8c9)
> 
> I get the error that this link is invalid or has been deleted.

It is working now with javascript enabled, google should fix there error messages.

> 
> [This summarises our thinking](https://gemini.google.com/share/843245d3d8c9)
I would not call this "summarising" - this is a fairly length bit of text.

To summarise the link: Avoiding the explicit `()` call can cause issues with `auto` and `templates` - which can be very confusing, and as such should be avoided. Did I get that right?

dschwoerer commented:
If so, you do not plan to do this in the future, and this PR is not a WIP and ready for review?
I asked 2 years ago https://github.com/boutproject/BOUT-dev/pull/2873#issuecomment-2011752571 - and would have started with a review if I would have known the answer.

Did you address https://github.com/boutproject/BOUT-dev/pull/2873#discussion_r1524926186 ?

If you get this updated, I will try to review this (and #3046 )

ZedThree committed:
Apply formatting

ZedThree committed:
Merge branch 'next' into refactor-coordinates

* next: (774 commits)
  Bump externalpackages/googletest from `73a63ea` to `d72f9c8`
  Bump pygments from 2.19.2 to 2.20.0
  Update pygithub requirement from ~=2.8 to ~=2.9
  Add docs for updating uv tools
  Update prek file
  Update uv.lock file
  Enable `uv lock`
  Add communicator for XZ planes
  Move `field_type()` to `FieldData`; deprecate `FieldData::is3D()`
  Apply `clang-tidy` fixes
  Communicate `FieldPerp` consistently with other Fields
  Revert "Use annotations instead of comments for `clang-tidy-review`"
  Fixes for prek
  Add headers from clang-tidy
  Add missing headers from clang-tidy
  Remove duplicate definitions
  Remove stray doc string
  Only generate grids once
  Expose Mesh::get for Field3D
  Formatting fixes
  ...

ZedThree committed:
Use single definition of `FieldMetric`

dschwoerer commented:
Now that https://github.com/boutproject/BOUT-dev/pull/3372 is merged, we can continue here :+1:

ZedThree commented:
Replaced by #3425
